### PR TITLE
OCPBUGS-12210: Prevent partially filled HPA behaviors from crashing kube-controller-manager

### PR DESCRIPTION
**Pull request by jkyros**
The short version here is that: 
- If you supply partial HPA behaviors (e.g. `ScaleUp` but not `ScaleDown`, etc ) in kube < 1.27, it will send the `kube-controller-manager` into `CrashLoopBackOff`
- This is fixed in kube 1.27+ by defaulting  to autoscaling v2: https://github.com/kubernetes/kubernetes/pull/114358 but we can't backport that type of change 
- So, since we're storing as v1 but consuming as v2 in the controller, we need to make sure that the behaviors aren't nil in the v2 object when someone creates or edits a v1 object to have partially filled behaviors

This PR: 
- ~~Defauts any nil behaviors when converting from v1 -> internal~~
- Makes the controller fill in missing behavior with defaults 
- Removes the "defaulter cheating" in the unit test that was masking the crash 
- Adds a test case to verify that it works 
- Is targeted straight to 4.13 because it's useless after that (if it were preexisting carry, it would have been dropped in 4.14) 

Updstream details: 

- I did [inquire upstream](https://github.com/kubernetes/kubernetes/issues/107038#issuecomment-1911327145) but we're already outside the n-3 supported versions and the fix is useless post 1.27,  so the "juice wasn't worth the squeeze". We still have at least one customer that needs this fixed so we'd just need to get this into 4.13 and 4.12 since they're still supported. 

Here is a straightforward crasher ( you might have to wait a little bit until the HPA touches it, but you should be able to see `kube-controller-manager` pods go into `CrashLoopBackoff`):

```
apiVersion: autoscaling/v1
kind: HorizontalPodAutoscaler
metadata:
  name: crasher
  namespace: test 
  labels:
    app: test
  annotations:
    autoscaling.alpha.kubernetes.io/behavior: '{"ScaleDown":{"StabilizationWindowSeconds":600,"SelectPolicy":"Max","Policies":[{"type":"Pods","value":1,"periodSeconds":1}]}}'
spec:
  scaleTargetRef:
    kind: Deployment
    name: test
    apiVersion: apps/v1
  minReplicas: 8
  maxReplicas: 25
  targetCPUUtilizationPercentage: 120
---
kind: Deployment
apiVersion: apps/v1
metadata:
  name: test
  namespace: test
  labels:
    app: test
spec:
  replicas: 2
  selector:
    matchLabels:
      app: test
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: test
    spec:
      containers:
        - resources:
            limits:
              cpu: 500m
              memory: 128Mi
            requests:
              cpu: 25m
              memory: 128Mi
          readinessProbe:
            httpGet:
              path: /
              port: 8080
              scheme: HTTP
            timeoutSeconds: 1
            periodSeconds: 10
            successThreshold: 1
            failureThreshold: 3
          terminationMessagePath: /dev/termination-log
          name: nginx
          livenessProbe:
            httpGet:
              path: /
              port: 8080
              scheme: HTTP
            timeoutSeconds: 1
            periodSeconds: 10
            successThreshold: 1
            failureThreshold: 3
          ports:
            - containerPort: 8080
              protocol: TCP
          imagePullPolicy: Always
          image: 'nginxinc/nginx-unprivileged:latest'
      restartPolicy: Always
      terminationGracePeriodSeconds: 30
      dnsPolicy: ClusterFirst
      securityContext: {}
      schedulerName: default-scheduler
  strategy:
    type: RollingUpdate
    rollingUpdate:
      maxUnavailable: 25%
      maxSurge: 25%
  revisionHistoryLimit: 10
  progressDeadlineSeconds: 600
``` 

Fixes: [OCPBUGS-12210]( https://issues.redhat.com/browse/OCPBUGS-12210)

### CODE REVIEW

**Comment by aravindhp** [comment]
/lgtm

I am fine with this approach

**Comment by deads2k** [comment]
the referenced upstream doesn't exist: 11358

**Comment by jkyros** [comment]
> the referenced upstream doesn't exist: 11358
Fixed, I'm a clown, I left out the 4.

File: pkg/apis/autoscaling/v1/conversion.go [hunk 1]
```diff
@@ -449,6 +450,20 @@ func Convert_v1_HorizontalPodAutoscaler_To_autoscaling_HorizontalPodAutoscaler(i
 	// drop round-tripping annotations after converting to internal
 	out.Annotations, _ = autoscaling.DropRoundTripHorizontalPodAutoscalerAnnotations(out.Annotations)
 
+	// Until kube 1.27 we're still storing autoscaling as v1, but the HPA controller is consuming it as v2. Behaviors are an annotation in v1 and can be partially
```
Comment (deads2k) [inline pkg/apis/autoscaling/v1/conversion.go:453]: what does this do to the annotations in question in a flow like
1. write object to API/etcd, scaleup gets set
2. read object as v1, do annotations make sense?

**Comment by deads2k** [comment]
where does the crash happen?  is it in conversion code?  if the failure happens in the kube-controller-manager, I'd rather see a kube-controller-manager patch because the blast radius is much smaller and doesn't have any persistent effect in the cluster.

**Comment by jkyros** [comment]
Eads and I talked over slack (thanks David).

 In summary, since the crash is in the controller, not the converter, we'd prefer to fix it in the controller to reduce the blast radius since this fix is only for two old versions. 

I've retooled this PR such that: 
- The fix is in the controller now
- It runs the HPA object through the behavior defaulters before we do any calculations on it (so the defaults are filled by the time it gets there) 
- I also added a test case for partial behaviors 
- I took out a "cheat" in the unit tests that was preventing us from finding the crash before now  

This does come with the side effect that if the controller modifies the object, it gets written back out with the missing values filled in, but that seems better to me than refusing to scale until someone tediously fills them in?

**Comment by joelsmith** [comment]
This looks a great low-risk solution for 4.13 & 4.12. Thanks!
/lgtm

**Comment by deads2k** [comment]
>This looks like a great low-risk solution for 4.13 & 4.12. Thanks!

Agreed, thank you

/approve

**Comment by jkyros** [comment]
@weinliu I know this isn't an actual "cherry pick" since we're stuffing it straight into 4.13, and it's not in a part of the repo that we "own", but would you be able to take a look here for the  `cherry-pick-approved` label? (Or at least sign off from the QE side so we can have someone add the label?). Thank you much!

**Comment by weinliu** [comment]
@sunilcio could you help?

**Comment by jkyros** [comment]
Thanks everyone! I'm only going to take this back one more to 4.12 (since it's EUS): 
/cherry-pick release-4.12

**Pull request merged**